POSIX_MAPPED_FILES and POSIX_MEMLOCK_RANGE select MMU based on CPU capability failing on platforms not enabling or using this capability

**Describe the bug**
Below tests are failing during building on platform **intel_ish_5_8_0**
libraries.libc.c11_threads.minimal
libraries.libc.c11_threads.picolibc
libraries.libc.c11_threads.picolibc.module
libraries.libc.c11_threads.picolibc.notls
libraries.libc.c11_threads.newlib
libraries.libc.c11_threads.newlib_nano 

This commit causes a regression [0ac931b](https://github.com/zephyrproject-rtos/zephyr/commit/0ac931b38647d875693a8b881fae0319692562a6)


**To Reproduce**
_example of one test from list:_
west twister -p intel_ish_5_8_0/intel_ish_5_8_0 -s libraries.libc.c11_threads.minimal


**Logs and console output**
```
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /zephyrproject/zephyr/twister-out/intel_ish_5_8_0_intel_ish_5_8_0/zephyr/tests/lib/c_lib/thrd/libraries.libc.c11_threads.minimal/zephyr/zephyr_pre0.map
:   <trimmed>
/zephyr-sdk-0.17.0/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/12.2.0/../../../../x86_64-zephyr-elf/bin/ld.bfd:
 zephyr/arch/arch/x86/core/libarch__x86__core.a(prep_c.c.obj): in function `z_prep_c':
/zephyrproject/zephyr/arch/x86/core/prep_c.c:40: undefined reference to `z_x86_mmu_init'

/zephyr-sdk-0.17.0/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/12.2.0/../../../../x86_64-zephyr-elf/bin/ld.bfd: zephyr/kernel/libkernel.a(mmu.c.obj):
in function `k_mem_map_phys_bare':
/zephyrproject/zephyr/kernel/mmu.c:961: undefined reference to `arch_mem_map'

collect2: error: ld returned 1 exit status
```

**Environment (please complete the following information):**

 - OS: Ubuntu
 - Toolchain:  zephyr-sdk-0.17.0
 -  Version used : v4.0.0-4669-g20516b78348d

> The posix layer does not select the mmu. It implies it, 

imply or select, it makes no difference if you should not be doing this in the first place. MMU shall only be selected by the implementation of MMU, i.e. X86_MMU or ARM_MMU, it should not be selected/implied by anyone else.

> but only when the architecture has an MMU  and when the application enables the memory mapping option group, as that particular hardware feature is required for functionality. 

Architecture might support MMU, but a platform might not be enabling that for various valid reasons. POSIX should not enable MMU in this case, it needs to just depend on it and if MMU is enabled on the platform, those posix features needing that would not be enabled. 

And of course, this is how MMU options are intended to work.

> And of course, this is how posix options are intended to work.

No, POSIX features requiring MMU shall not enable MMU, it shall depend on it instead.

> However, the main issue here is that the x86 architecture has an inconsistent implementation of this particular aspect of the arch interface (also mentioned in the proposed fix), which is least invasive.

What is the inconsistency here exactly? I tried to explain above how this is supposed to work and this is not an X86 issue, this is how MMU on all architectures is working now , this is how MPU works and this is how it was since this was introduced.

try:

`west build -p -b qemu_cortex_a53  tests/benchmarks/posix/threads -T benchmark.posix.threads  -- -DCONFIG_ARM_MMU=n`

Which represents an arch that has MMU support but does not want to enable MMU, this could be a platform configuration, similar to the x86 nommu variant, this exposes exactly the same issue seen on ish:

POSIX enables MMU, although the platform does not enable or want that, we end up with:

```
-- Zephyr version: 4.0.99 (/home/anashif/zephyrproject/zephyr), build: v4.0.0-5519-gd339b961321f
[101/166] Building C object zephyr/subsys/demand_paging/eviction/CMakeFiles/subsys__demand_paging__eviction.dir/lru.c.obj
FAILED: zephyr/subsys/demand_paging/eviction/CMakeFiles/subsys__demand_paging__eviction.dir/lru.c.obj
ccache /home/anashif/zephyr-sdk-0.17.0/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=1024 -DPICOLIBC_DOUBLE_PRINTF_SCANF -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/anashif/zephyrproject/zephyr/kernel/include -I/home/anashif/zephyrproject/zephyr/arch/arm64/include -I/home/anashif/zephyrproject/zephyr/build/zephyr/include/generated/zephyr -I/home/anashif/zephyrproject/zephyr/include -I/home/anashif/zephyrproject/zephyr/build/zephyr/include/generated -I/home/anashif/zephyrproject/zephyr/soc/arm -I/home/anashif/zephyrproject/zephyr/include/zephyr/posix -I/home/anashif/zephyrproject/zephyr/soc/arm/qemu_cortex_a53/. -I/home/anashif/zephyrproject/zephyr/subsys/testsuite/include -isystem /home/anashif/zephyrproject/zephyr/lib/libc/common/include -Wshadow -fno-strict-aliasing -Os -imacros /home/anashif/zephyrproject/zephyr/build/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-a53 -mabi=lp64 --sysroot=/home/anashif/zephyr-sdk-0.17.0/aarch64-zephyr-elf/aarch64-zephyr-elf -imacros /home/anashif/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/anashif/zephyrproject/zephyr/tests/benchmarks/posix/threads=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/anashif/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/anashif/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -moverride=tune=no_ldp_stp_qregs -specs=picolibc.specs -D_POSIX_THREADS -std=c99 -MD -MT zephyr/subsys/demand_paging/eviction/CMakeFiles/subsys__demand_paging__eviction.dir/lru.c.obj -MF zephyr/subsys/demand_paging/eviction/CMakeFiles/subsys__demand_paging__eviction.dir/lru.c.obj.d -o zephyr/subsys/demand_paging/eviction/CMakeFiles/subsys__demand_paging__eviction.dir/lru.c.obj -c /home/anashif/zephyrproject/zephyr/subsys/demand_paging/eviction/lru.c
/home/anashif/zephyrproject/zephyr/subsys/demand_paging/eviction/lru.c: In function 'k_mem_paging_eviction_select':
/home/anashif/zephyrproject/zephyr/subsys/demand_paging/eviction/lru.c:174:32: error: 'ARCH_DATA_PAGE_DIRTY' undeclared (first use in this function)
  174 |         *dirty_ptr = ((flags & ARCH_DATA_PAGE_DIRTY) != 0);
      |                                ^~~~~~~~~~~~~~~~~~~~
/home/anashif/zephyrproject/zephyr/subsys/demand_paging/eviction/lru.c:174:32: note: each undeclared identifier is reported only once for each function it appears in
[110/166] Building C object zephyr/subsys/demand_paging/backing_store/CMakeFiles/subsys__demand_paging__backing_store.dir/ram.c.obj
FAILED: zephyr/subsys/demand_paging/backing_store/CMakeFiles/subsys__demand_paging__backing_store.dir/ram.c.obj
ccache /home/anashif/zephyr-sdk-0.17.0/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=1024 -DPICOLIBC_DOUBLE_PRINTF_SCANF -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/anashif/zephyrproject/zephyr/kernel/include -I/home/anashif/zephyrproject/zephyr/arch/arm64/include -I/home/anashif/zephyrproject/zephyr/build/zephyr/include/generated/zephyr -I/home/anashif/zephyrproject/zephyr/include -I/home/anashif/zephyrproject/zephyr/build/zephyr/include/generated -I/home/anashif/zephyrproject/zephyr/soc/arm -I/home/anashif/zephyrproject/zephyr/include/zephyr/posix -I/home/anashif/zephyrproject/zephyr/soc/arm/qemu_cortex_a53/. -I/home/anashif/zephyrproject/zephyr/subsys/testsuite/include -isystem /home/anashif/zephyrproject/zephyr/lib/libc/common/include -Wshadow -fno-strict-aliasing -Os -imacros /home/anashif/zephyrproject/zephyr/build/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-a53 -mabi=lp64 --sysroot=/home/anashif/zephyr-sdk-0.17.0/aarch64-zephyr-elf/aarch64-zephyr-elf -imacros /home/anashif/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/anashif/zephyrproject/zephyr/tests/benchmarks/posix/threads=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/anashif/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/anashif/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -moverride=tune=no_ldp_stp_qregs -specs=picolibc.specs -D_POSIX_THREADS -std=c99 -MD -MT zephyr/subsys/demand_paging/backing_store/CMakeFiles/subsys__demand_paging__backing_store.dir/ram.c.obj -MF zephyr/subsys/demand_paging/backing_store/CMakeFiles/subsys__demand_paging__backing_store.dir/ram.c.obj.d -o zephyr/subsys/demand_paging/backing_store/CMakeFiles/subsys__demand_paging__backing_store.dir/ram.c.obj -c /home/anashif/zephyrproject/zephyr/subsys/demand_paging/backing_store/ram.c
/home/anashif/zephyrproject/zephyr/subsys/demand_paging/backing_store/ram.c: In function 'k_mem_paging_backing_store_page_finalize':
/home/anashif/zephyrproject/zephyr/subsys/demand_paging/backing_store/ram.c:132:25: error: 'ARCH_UNPAGED_ANON_ZERO' undeclared (first use in this function)
  132 |         if (location == ARCH_UNPAGED_ANON_ZERO || location == ARCH_UNPAGED_ANON_UNINIT) {
      |                         ^~~~~~~~~~~~~~~~~~~~~~
/home/anashif/zephyrproject/zephyr/subsys/demand_paging/backing_store/ram.c:132:25: note: each undeclared identifier is reported only once for each function it appears in
/home/anashif/zephyrproject/zephyr/subsys/demand_paging/backing_store/ram.c:132:63: error: 'ARCH_UNPAGED_ANON_UNINIT' undeclared (first use in this function)
  132 |         if (location == ARCH_UNPAGED_ANON_ZERO || location == ARCH_UNPAGED_ANON_UNINIT) {
      |                                                               ^~~~~~~~~~~~~~~~~~~~~~~~
[124/166] Building C object zephyr/lib/posix/options/CMakeFiles/lib__posix__options.dir/pthread.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/anashif/zephyrproject/zephyr/build
```

not enabling MMU is a valid configuration/option, i.e. hello world works fine on this plaform without MMU:

`west build -t run  -p -b qemu_cortex_a53  samples/hello_world/ -- -DCONFIG_ARM_MMU=n`

```
*** Booting Zephyr OS build v4.0.0-5519-gd339b961321f ***
Hello World! qemu_cortex_a53/qemu_cortex_a53
qemu-system-aarch64: terminating on signal 2
```

--- a/lib/posix/options/Kconfig.mem
+++ b/lib/posix/options/Kconfig.mem
@@ -27,10 +27,11 @@ config POSIX_SHARED_MEMORY_OBJECTS
 
 config POSIX_MAPPED_FILES
 	bool "POSIX memory-mapped files"
-	# disable Xtensa for now until linker issues are resolved
-	imply MMU if (CPU_HAS_MMU && !XTENSA)
 	help
 	  Select 'y' here and Zephyr will provide support for mmap(), msync(), and munmap().
+
+	  Note: This feature depends on hardware MMU support. If the underlying platform does not
+	  support an MMU, then affected POSIX API functions may return -1 and set errno to ENOTSUP.
 
 	  For more information, please see
 	  https://pubs.opengroup.org/onlinepubs/9699919799.orig/functions/V2_chap02.html
@@ -43,16 +44,22 @@ config POSIX_MEMLOCK
 	help
 	  Select 'y' here and Zephyr will provide support for mlockall() and munlockall().
 
+	  Note: This feature depends on hardware MMU support as well as DEMAND_PAGING. If the
+	  underlying platform does not support an MMU or DEMAND_PAGING, then affected POSIX API
+	  functions will return -1 and set errno to ENOTSUP.
+
 	  For more information, please see
 	  https://pubs.opengroup.org/onlinepubs/9699919799.orig/functions/V2_chap02.html
 	  https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_port.html#tag_24_03_04
 
 config POSIX_MEMLOCK_RANGE
 	bool "POSIX range memory locking"
-	imply MMU if (CPU_HAS_MMU && ARCH_HAS_DEMAND_PAGING)
 	imply DEMAND_PAGING
 	help
 	  Select 'y' here and Zephyr will provide support for mlock() and munlock().
+
+	  Note: This feature depends on hardware MMU support. If the underlying platform does not
+	  support an MMU, then affected POSIX API functions will return -1 and set errno to ENOTSUP.
 
 	  For more information, please see
 	  https://pubs.opengroup.org/onlinepubs/9699919799.orig/functions/V2_chap02.html

--- a/tests/posix/xsi_realtime/testcase.yaml
+++ b/tests/posix/xsi_realtime/testcase.yaml
@@ -13,10 +13,6 @@ common:
   platform_exclude:
     # linker_zephyr_pre0.cmd:140: syntax error (??)
     - qemu_xtensa/dc233c
-    # CONFIG_MMU=y but no arch_mem_map() or arch_mem_unmap()
-    - intel_ish_5_4_1
-    - intel_ish_5_6_0
-    - intel_ish_5_8_0
     - native_sim
     - native_sim/native/64
 tests:

